src/pymmcore_widgets/hcs/_fov_widget/_well_graphics_view.py

```diff
@@ -0,0 +1,178 @@
+        items = self.scene().items(scene_pos)
+        for item in items:
+            if idx := item.data(DATA_POSITION_INDEX):
+                self.positionClicked.emit(idx, item.data(DATA_POSITION))
+                break
+
+    def setWellSize(self, width_mm: float | None, height_mm: float | None) -> None:
+        """Set the well size width and height in mm."""
+        self._well_width_um = (width_mm * 1000) if width_mm else None
+        self._well_height_um = (height_mm * 1000) if height_mm else None
+
+    def _well_rect(self) -> QRectF:
+        """Return the QRectF of the well area."""
+        if not self._well_width_um or not self._well_height_um:
+            return QRectF()
+        return QRectF(
+            -self._well_width_um / 2,
+            -self._well_height_um / 2,
+            self._well_width_um,
+            self._well_height_um,
+        )
+
+    def _draw_outline(self) -> None:
+        """Draw the outline of the well area."""
+        if self._outline_item:
+            self._scene.removeItem(self._outline_item)
+
+        pen = QPen(QColor(Qt.GlobalColor.green))
+        pen.setWidth(self._scaled_pen_size())
+        if self._is_circular:
+            self._outline_item = self._scene.addEllipse(self._well_rect(), pen=pen)
+        else:
+            self._outline_item = self._scene.addRect(self._well_rect(), pen=pen)
+
+    def _draw_fovs(self, plan: useq.RelativeMultiPointPlan) -> None:
+        """Draw the fovs in the scene as rectangles."""
+        # delete existing FOVs
+        while self._fov_items:
+            self._scene.removeItem(self._fov_items.pop())
+
+        half_fov_width = self._fov_width_um / 2
+        half_fov_height = self._fov_height_um / 2
+
+        # constrain random points to our own well size, regardless of the plan settings
+        # TODO: emit a warning here?
+        if isinstance(plan, useq.RandomPoints):
+            kwargs = {}
+            if self._well_width_um:
+                kwargs["max_width"] = self._well_width_um - half_fov_width * 1.4
+            if self._well_height_um:
+                kwargs["max_height"] = self._well_height_um - half_fov_height * 1.4
+            plan = plan.replace(**kwargs)
+
+        pen = QPen(Qt.GlobalColor.white)
+        pen.setWidth(self._scaled_pen_size())
+        line_pen = QPen(QColor(0, 0, 0, 100))
+        line_pen.setWidth(int(self._scaled_pen_size() // 1.5))
+
+        # iterate over the plan greedily, catching any warnings
+        # and then alert the model if we didn't get all the points
+        # XXX: I'm not sure about this pattern... feels like the model should be
+        # able to handle this itself, but this is perhaps the only place we actually
+        # iterate over the plan
+        with warnings.catch_warnings():
+            warnings.simplefilter("ignore")
+            points = list(plan)
+        if len(points) < getattr(plan, "num_points", 0):
+            self.maxPointsDetected.emit(len(points))
```

Comment:
The `QGraphicScene` coordinate are inverted compared to the one we use in `useq` so maybe it is worth adding this here

```suggestion
            self.maxPointsDetected.emit(len(points))

        # invert the y axis just for visualization purposes since point are calculated
        # in a coordinate system where the y axis is +up -down. In QGraphicScene the
        # y axis is +down -up.
        points = [p.replace(x=p.x, y=-p.y) for p in points]
        
```
We go form this:
<img width="1014" alt="Screenshot 2024-07-09 at 9 52 47 AM" src="https://github.com/pymmcore-plus/pymmcore-widgets/assets/70725613/0a8549c7-32c0-4d5e-a646-05ae1a9151ea">
to this:
<img width="1014" alt="Screenshot 2024-07-09 at 12 11 01 PM" src="https://github.com/pymmcore-plus/pymmcore-widgets/assets/70725613/c4147845-63b9-4e1b-9ad6-b5dbfd610ee5">

Reply:
good catch.  
btw: in general we use the `.replace` method that I added in useq schema too much.  it was meant to be a public convenience for mutating otherwise "frozen" items, with a) a slightly nicer api than the already-provided `model_copy` function that pydantic provides  and b) ensured validation (it actually dumps and recreates the full object).  However, if you *know* that the value you are replacing it with is valid, then you shouldn't use `.replace`, you should use `model_copy` 👍